src/backend/Level0/CHIPBackendLevel0.cc

```diff
@@ -1296,35 +1257,36 @@ CHIPEvent *CHIPQueueLevel0::enqueueMarkerImpl() {
   // Done via GET_COMMAND_LIST
   auto Status = zeCommandListAppendSignalEvent(
       CommandList,
-      MarkerEvent->get("MarkerEvent: zeCommandListAppendSignalEvent"));
+      std::static_pointer_cast<CHIPEventLevel0>(MarkerEvent)->peek());
   CHIPERR_CHECK_LOG_AND_THROW(Status, ZE_RESULT_SUCCESS, hipErrorTbd);
   executeCommandList(CommandList);
 
   return MarkerEvent;
 }
 
-CHIPEvent *
-CHIPQueueLevel0::enqueueBarrierImpl(std::vector<CHIPEvent *> *EventsToWaitFor) {
-  CHIPEventLevel0 *EventToSignal =
-      (CHIPEventLevel0 *)Backend->createCHIPEvent(ChipContext_);
-
-  EventToSignal->sanityCheck();
+std::shared_ptr<CHIPEvent> CHIPQueueLevel0::enqueueBarrierImpl(
+    std::vector<std::shared_ptr<CHIPEvent>> EventsToWaitFor) {
+  std::shared_ptr<CHIPEvent> EventToSignal =
+      static_cast<CHIPBackendLevel0 *>(Backend)->createCHIPEvent(ChipContext_);
   EventToSignal->Msg = "barrier";
   size_t NumEventsToWaitFor = 0;
 
-  NumEventsToWaitFor = EventsToWaitFor ? EventsToWaitFor->size() : 0;
+  NumEventsToWaitFor = EventsToWaitFor.size();
 
   ze_event_handle_t *EventHandles = nullptr;
   ze_event_handle_t SignalEventHandle = nullptr;
 
-  SignalEventHandle = ((CHIPEventLevel0 *)(EventToSignal))->peek();
+  SignalEventHandle =
+      std::static_pointer_cast<CHIPEventLevel0>(EventToSignal)->peek();
 
   if (NumEventsToWaitFor > 0) {
     EventHandles = new ze_event_handle_t[NumEventsToWaitFor];
     for (size_t i = 0; i < NumEventsToWaitFor; i++) {
-      CHIPEventLevel0 *ChipEventLz = (CHIPEventLevel0 *)(*EventsToWaitFor)[i];
+      std::shared_ptr<CHIPEvent> ChipEvent = EventsToWaitFor[i];
+      std::shared_ptr<CHIPEventLevel0> ChipEventLz =
+          std::dynamic_pointer_cast<CHIPEventLevel0>(ChipEvent);
```

Comment:
Unless i'm mistaken, this will introduce requirement to compile the code with RTTI.

Reply:
typo

Reply:
typo. Fixed.